src/docfx/config/ops/CredentialHandler.cs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Identity;
+using Azure.Security.KeyVault.Secrets;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CredentialHandler : DelegatingHandler
+    {
+        private const string DocsOPSTokenHeader = "X-OP-BuildUserToken";
+
+        public static readonly DocsEnvironment DocsEnvironment = GetDocsEnvironment();
+
+        private readonly Action<HttpRequestMessage>? _credentialProvider;
+
+        // TODO: use Azure front door endpoint when it is stable
+        private static readonly string s_docsProdServiceEndpoint =
+            Environment.GetEnvironmentVariable("DOCS_PROD_SERVICE_ENDPOINT") ?? "https://op-build-prod.azurewebsites.net";
+
+        private static readonly string s_docsPPEServiceEndpoint =
+            Environment.GetEnvironmentVariable("DOCS_PPE_SERVICE_ENDPOINT") ?? "https://op-build-sandbox2.azurewebsites.net";
+
+        private static readonly string s_docsInternalServiceEndpoint =
+            Environment.GetEnvironmentVariable("DOCS_INTERNAL_SERVICE_ENDPOINT") ?? "https://op-build-internal.azurewebsites.net";
+
+        private static readonly string s_docsPerfServiceEndpoint =
+            Environment.GetEnvironmentVariable("DOCS_PERF_SERVICE_ENDPOINT") ?? "https://op-build-perf.azurewebsites.net";
+
+        private static readonly SecretClient s_secretClient = new(new("https://docfx.vault.azure.net"), new DefaultAzureCredential());
+
+        private static readonly Lazy<Task<string>> s_opsTokenProd = new(() => GetSecret("OpsBuildTokenProd"));
+        private static readonly Lazy<Task<string>> s_opsTokenSandbox = new(() => GetSecret("OpsBuildTokenSandbox"));
+
+        public CredentialHandler(Action<HttpRequestMessage>? credentialProvider, HttpMessageHandler innerHandler)
+            : base(innerHandler)
+        {
+            _credentialProvider = credentialProvider;
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            using (PerfScope.Start($"[{nameof(OpsConfigAdapter)}] Executing request '{request.Method} {request.RequestUri}'"))
+            {
+                _credentialProvider?.Invoke(request);
+                if (TryExtractDocsEnvironmentFromUrl(request.RequestUri?.AbsoluteUri, out _))
+                {
+                    return await HandleOpsRequest(request, cancellationToken);
+                }
+
+                return await base.SendAsync(request, cancellationToken);
+            }
+        }
+
+        public static string BuildServiceEndpoint(DocsEnvironment? environment = null)
```

Comment:
>BuildServiceEndpoint [](start = 29, length = 20)

It is kind of weird to me that `CredentialManager` exposes a method called `BuildServiceEndpoint`. Shall we consider extracting environment and endpoint info into a separate class, e.g. named `OpsEnvironments`?

Reply:
moved back to opsAccessor